### ISSUE TITLE
Revamping app data

### ISSUE THREAD

**Issue by vkgnosis** [cowprotocol/services#1465]
## Explanation of app data

App data is extra information that is associated with orders. This information has no semantics in the smart contract except that it is signed by order creator. App data is useful to other parts of the system. Examples:

- You specify that you want your order to be COW only. This has meaning only to the backend services. They treat your order differently during auctions. The smart contract cannot protect you against having this request violated. However, you can prove that you signed your order with this intention and complain to the backend for not upholding it. Likewise the backend can prove that the order was signed this way to explain why it had special treated for the order.
- You use the official frontend, which signals through app data what your intended slippage was, that you selected in the UI. This can later be used for debugging.

On the smart contract level app data is freely choosable 32 bytes of signed order data. This is a low amount of space. In order to refer to truly arbitrary data, we interpret those bytes as a cryptographic hash of the full data. The full app data  is thus signed by the user when they sign the order.

The words "app data" can refer to app data on a different levels. To disambiguate I am using the following terms:
- "contract app data" is the 32 bytes app data field in the contract's order struct.
- "frontend app data" is app data that has meaning to the frontend. It has a [schema](https://github.com/cowprotocol/app-data).
- "full app data" is the document that is hashed to set the contract app data. This concept is newly introduced in this text, below.
- "backend app data" is app data that has meaning to the backend. For example the hypothetical "cow only" flag from above.

Our only use case for contract app data used to be frontend app data. The frontend has a schema that frontend app data follows and a way to turn frontend app data into a 32 byte hash. The hash is used for contract app data and doubles as an IPFS CID. This allows app data to be stored and retrieved through IPFS. The backend used to not concern itself with app data and treated contract app data as bytes with no meaning.

## New requirements

Recently it has become clear that we need backend app data. I already mentioned the hypothetical "cow only" flag. A real use case we soon want to have is to contain [custom interactions](https://github.com/cowprotocol/services/issues/1457). 

We could specify backend app data through new parameters in the "create order" API but then it wouldn't be signed. Or the user would have to sign multiple times when creating an order. To avoid this, we need to go through contract app data.

## Full app data

Contract app data is already used by frontend app data. We don't want to break the frontend's use case so we need to combine both forms of app data. This is full app data.

Full app data is an arbitrary json object, that is sent to the backend on order creation. It contains any information an api user wants to attach to the order like frontend app data. It also contains a reserved field specifically for backend app data. The backend validates and stores full app data in the backend database so that it can be produced again later.

The method of turning full app data into contract app data is completely specified and deterministic. This allow later validation of full app data. The hashing scheme is chosen in a way that is compatible with IPFS. This allows app data to be easily stored on IPFS (but it is also stored in the backend database).

## Specification

The create order API changes to accommodate full app data. Instead of passing in contract app data (the existing `appData` parameter), you pass in full app data. Full app data is a json string. The string contains a utf-8 encoded json object. This json object models full app data. Contract app data is implicitly (no longer explicitly passed in the API)  assumed to be the keccak-256 hash of the utf-8 encoded bytes of the full app data string.

The full app data json object can be arbitrarily chosen. For example, you could follow the frontend's app data schema. The one exception to this is that the key `backend` is reserved for backend app data. This key cannot be arbitrarily chosen because it must conform to the backend app data's schema and is validated by the backend. Currently there is no backend app data and the key is omitted.

There are some details of the new API that I'm not fully specifying here like the name of the new full app data parameter or that for backward compatibility will keep accepting the old create order parameters for some time.

### Example

You want to set this full app data json:
```json
{
    "hello": "world"
}
```

You encode this json object into a string. This could result in the minimized string `{"hello":"world"}`. You keccak-256 hash the utf-8 bytes of this string. This hash is what you set the contract app data to when signing your order. You call the create order API. In the POST json body, you set the field `appData` to the string that you signed.

On the backend side the same hashing scheme is used to arrive at the same contract app data. If either side made a mistake during hashing, then signature verification fails. The backend stores the full app data string in the backend database.

### IPFS

From the backend's point of view IPFS is not required. The backend already stores full app data on order creation. To support the frontend's current workflow and ease decentralization of app data in the future, we define scheme of turning full app data into IPFS CIDs.

Assuming that you have the keccak-256 hash of the full app data there is a canonical way of turning it into a CID that refers to the full app data that resulted in the hash. This allows you to go from contract app data (which is the keccak-256 hash) to CID and retrieve the full app data that was used to create the order from IPFS (assuming it was uploaded to IPFS under this canonical scheme). The canonical way to create the CID is as follows:

Create a CIDv1 with codec RAW (0x55), hash algorithm keccak-256 (0x1b), hash length 32, hash content contract app data.

Here is equivalent Rust code:

```rust
/// The return value is the raw bytes of the CID. It is not multibase encoded.
pub fn create_ipfs_cid(app_data_hash: &[u8; 32]) -> [u8; 36] {
    let mut cid = [0u8; 4 + 32];
    cid[0] = 1; // cid version
    cid[1] = 0x55; // raw codec
    cid[2] = 0x1b; // keccak hash algorithm
    cid[3] = 32; // keccak hash length
    cid[4..].copy_from_slice(app_data_hash);
    cid
}
```

This is a CID version 1, that refers to a raw IPFS block that was hashed with keccak-256. To provide such a block to IPFS, you could use the IPFS command line:

```sh
~ echo '{"appCode":"CoW Swap","environment":"production","metadata":{"quote":{"slippageBips":"50","version":"0.2.0"},"orderClass":{"orderClass":"market","version":"0.1.0"}},"version":"0.6.0"}' | ipfs block put --mhtype keccak-256
bafkrwiek6tumtfzvo6yivqq5c7jtdkw6q3ar5pgfcjdujvrbzkbwl3eueq
```

## Implementation 

The backend already accepts and stores full app data as specified in the openapi docs. It also accepts the old contract app data to be fully backward compatible. Eventually we want to get rid of the old app data. To achieve that we need to do several things:
- Update frontend code that places order to follow the new scheme.
- Update documentation like [this](https://docs.cow.fi/front-end/creating-app-ids) with the new format.
- Get important API users to follow the new scheme.
- Update our code that reads app data from IPFs to try the old and the new scheme because either might have been used to put app data in IPFS. Alternatively we can ingest all existing app data into the backend and make app data reading code always go through a backend API. This makes IPFS completely optional.

**Comment by MartinquaXD** [comment]
Really nice write up! I think I have to go over it a few more times to catch the nuances but here is an initial thought:
> If "full app data" is not set, everything works as before. If it is set then the backend validates that "app data" is set to the keccak256 hash of the "full app data" contents.


This seems problematic in terms of accountability. If for example a user were to use the `appdata` as it was initially intended (create JSON, upload to IPFS, store hash as appdata) but they **don't** set `full_app_data` we would not do any validations and treat the `appdata` as random meaningless bytes. If the user were then to complain that we didn't respect the hypothetical `cow_only` flag in their IPFS document that provably matches the specified `appdata` we would not have stored enough information to prove that they didn't send the `full_app_data` which would be required to legitimately ignore the contents of `appdata`.
Long story short: I think if we start to encode data relevant to the order execution in the `appdata` we should immediately enforce that non-zero `appdata` always comes with the `full_app_data`.

> Eventually we should make "full app data" mandatory or enforce that "app data" is all zero if "full app data" isn't set. This gives us more certainty that we didn't lose a user's full app data. However, this is a breaking change so we shouldn't do it immediately.

It's only a breaking change if we modify the existing endpoint for it. We could also consider creating `POST /api/v2/orders` although I think even having this as a breaking change is probably no big deal.

**Comment by vkgnosis** [comment]
> This seems problematic in terms of accountability. 

At the moment there is no backend relevant app data. So the user complaint in the example doesn't make sense to me. It is true though that as long as we keep this backward compatibility, we cannot prove that a user really signed empty app data instead of the backend maliciously ignoring app data. This is the case still with a v2 route if we don't disable the v1 route. The reason I think backward compatibility is important for some time is that we'd break any direct API user that is setting app data for whatever reason.

**Comment by MartinquaXD** [comment]
> At the moment there is no backend relevant app data. So the user complaint in the example doesn't make sense to me.

It's true that we **currently** don't store relevant data but my point is that these complaints could theoretically happen for the entire period between `starting to store relevant data` and `enforcing strict verification of app_data`.

**Comment by anxolin** [comment]
Thanks for the writeup. I think it helpsto understand the proposal. 

> In https://github.com/cowprotocol/services/issues/1457 we noticed that the IPFS approach isn't useful and that it makes more sense to handle app data in the backend.

I'm not sure I follow this. Do you mean, that is not useful because you would need to fetch the IPFS document and this introduces some asynchronous logic in the order processing? 

If the point is that to make it simpler and faster the backend prefers to receive the document information together with the hash, I think is a fair ask, I also think implementation will be easier in the backend to be able to classify and decorate orders synchronously.  However, I just wouldn't say is "not useful".

> It is a string because the hash depends on the exact json encoding, for which there exists no canonical method of creation, so we must fall back to the api user choosing the encoding.

Makes sense to me. This way there's no discussions on how to encode the JSON and the app-data will match. 


> If "full app data" is not set, everything works as before. If it is set then the backend validates that "app data" is set to the keccak256 hash of the "full app data" contents. The backend performs further validation on the field. For example, the backend checks that it is valid json, that the backend specific app data has the expected format, that the app data as a whole is under some size limit. If any validation fails, the order is rejected.

Agreed with most of this. I think is a great proposal. Probably I would agree with @MartinquaXD it makes sense to make a hard cut between the old way and the new way, and use `v2` 

The part I don't agree is in using `keccak256` for the validation. We should use sha-256 hashing IMO, we should follow  https://docs.ipfs.tech/concepts/content-addressing/ and make it a CID. This will give us hashes that are a digest of the content, and will allow us to have it also in IPFS which helps with the decentralization of the protocol.

i.e. https://github.com/multiformats/js-cid

```bash
const bytes = new TextEncoder('utf8').encode('OMG!')
const hash = await multihashing(bytes, 'sha2-256')
const cid = new CID(1, 'dag-pb', hash)
```

For JS/TS projects, they should use https://github.com/cowprotocol/app-data, there's some utils there to handle this. 

> The mentioned backend specific app data is a reserved field in the full app data top level json object. The field is called "reserved". The contents of this field have special meaning to the backend and how it handles the order. Other top level fields can be freely chosen.

I got lost a bit in this part. I would suggest doing the specification in Swagger

> new route GET /api/v1/app_data/<order_uid>:
> The route returns the contents of the "full app data" used to create this order. This can be used to inspect the app data and validate the hash. It is important that the body of the response is the "full app data" field (a string) exactly. Otherwise the hash cannot be verified.

I would suggest `/orders/<order_uid>/full-app-data` 

- Reason, to nest it in the order is because it's an order field, so its part of this resource
- to use hiphen, because the contract field is camelCase (`appData`) and https://stackoverflow.com/questions/10302179/hyphen-underscore-or-camelcase-as-word-delimiter-in-uris#:~:text=The%20standard%20best%20practice%20for,API%20Design%20Rulebook%22%20from%20Oreilly. 
- I suggest calling it `full-app-data` and not `app-data` , because the later is a 32 bytes hash and the former a string.


I would also suggest, if possible to define that this endpoint will take care of the mime-type, and return JSON most of the time (when the appData is actually a JSON, which is the normal case, and i understood it will even be enforced).

If you want to offer text mime-types for non-JSON content is fine, we will handle that too, but I think is less interesting.

**Comment by anxolin** [comment]
Regarding where to store the string, I would think is not too expensive to store in the Database, and seems easier. In any case, I would really make sure we also post it to Pinata to keep it decentralised and keep the current behaviour. 

I hope this makes things easier: 

https://docs.rs/pinata-sdk/latest/pinata_sdk

**Comment by vkgnosis** [comment]
> > In #1457 we noticed that the IPFS approach isn't useful and that it makes more sense to handle app data in the backend.
> 
> I'm not sure I follow this. Do you mean, that is not useful because you would need to fetch the IPFS document and this introduces some asynchronous logic in the order processing?

Until now the backend did not care about full app data. Now (with custom interactions) the backend definitely needs to store a copy of full app data, because otherwise the full app data could disappear at any moment but we need it to prove that the user really signed it. This is why I said it isn't useful (to the backend).

> The part I don't agree is in using keccak256 for the validation. 

I arbitrarily picked keccak because it is used by ethereum. Your suggestion of using an IPFS compatible hashing scheme is interesting. Then full app data could be looked up either on IPFS or through the backend.

On the other hand I'm unsure whether this adds meaningful decentralization. By virtue of being a cryptographic hash and the backend having an api to retrieve it, anyone could with this proposal create a backup of the full app data. They would be able to prove that the full app data matches some order through the hash. This doesn't have to be done through IPFS. But we could standardize to IPFS because it already exists and is widely used. In that case the the frontend (or any other API user) would provide the full app data to the backend when creating an order and could also upload to IPFS.

On the other hand the IPFS part is not relevant to the backend. We would be picking a more complicated hashing scheme just to support this use case, which anyone using the API would have to copy. And I don't know whether there is a real advantage of supporting it. I'm not saying that there isn't. I really don't know.

**Comment by vkgnosis** [comment]
I'm convinced. Let's do IPFS compatible CID. Just need to research how to implement it in Rust.

**Comment by alfetopito** [comment]
> I'm convinced. Let's do IPFS compatible CID. Just need to research how to implement it in Rust.

Alex (Hermann) had done it in the past, and it's what I based the TS implementation on.
Probably @bh2smith would know where to find it since I believe he worked on the same codebase for the affiliate endpoint.
Right, Ben?

**Comment by bh2smith** [comment]
> Probably @bh2smith would know where to find it since I believe he worked on the same codebase for the affiliate endpoint.


I know where the rust implementation is... somewhere in and around right here:

https://github.com/cowprotocol/dune-bridge/blob/main/src/models/app_data_json_format.rs

**Comment by vkgnosis** [comment]
Code linked above has Rust models of the app data json but not code to calculate the CID. The CID is unfortunately more complicated than expected because it isn't just a hash over the app data. IPFS has several ways files can be hashed. The simplest form is called raw (see [this](https://github.com/multiformats/multicodec/blob/master/table.csv)) but the one used by the frontend is dag-pb. In raw we would just hash the content but in dag-pb . dag-pb is made for encoding several files (like a tar or zip file) so it is more complicated. See [this](https://ipld.io/specs/codecs/dag-pb/spec/) and [this](https://github.com/ipfs/js-ipfs-unixfs). It would have been to have done "raw" all along but in order to keep backward compatibility we can't change it.

**Comment by bh2smith** [comment]
>  The CID is unfortunately more complicated than expected because it isn't just a hash over the app data. IPFS has several ways files can be hashed

This is also contained in that project I can find it for you. It is DEFINITELY a non-trivial task, but I guarantee the code is in there.

**Comment by vkgnosis** [comment]
Thanks Ben. I think this code calculates a CID from the smart contract order app data field. This allows you to look up the full app data on ipfs from the app data field. The thing I was talking about is having the full app data json and calculating the CID from that. I don't think that happens in that project (or I missed it again, sorry).

**Comment by bh2smith** [comment]
> The thing I was talking about is having the full app data json and calculating the CID from that. I don't think that happens in that project (or I missed it again, sorry).

No that is not part of this project. I believe the code for that is only available in our front end (i.e. typescript somehwere). I htink its also in docs.cow.fi

somewhere around here might have a link to the code used to turn the json into CID and/or app hash.

https://docs.cow.fi/front-end/creating-app-ids/create-the-order-meta-data-file

**Comment by anxolin** [comment]
@vkgnosis not sure if this is valid suggestion, but what about https://crates.io/crates/cid

**Comment by anxolin** [comment]
Also, does this chat GPT help? 


https://github.com/cowprotocol/services/assets/2352112/90dbfb06-2525-4c59-8355-fb31ea0724e7

**Comment by cowfee** [comment]
I would just call the field "backend" instead of "reserved" to make the purpose clearer. You can also have an optional "reserved" field that can be used for anything else in the future if needed. Of course the "backend" field is also reserved for use only by the backend and this should be clearly documented.

**Comment by anxolin** [comment]
@vkgnosis thanks for such a deep dive into CIDs

I'm not sure I understood fully your proposal. 
Let me re-phrase to see if I understand all. 

I get you are not happy with our default. Right now we assume `sha2-256` and `MerkleDAG protobuf` as it's what IPFS was using for their links. One of the reasons is that the library might change the "defaults". I think, in this case, we can keep using the old ones (they are stable, and we can generate cidv1 codes forever using this multihash).
<img width="1364" alt="image" src="https://github.com/cowprotocol/services/assets/2352112/eeffae6b-5e94-467a-a153-fd35e0136799">

- You are suggesting changing the format we have now, instead of being the hash pointing to the IPFS document, you suggest saving only the content hash, and then uploading to IPFS and having a way to derive the IPFS cid from the content hash (assuming a hashing algorithm, codec, etc)
- You argue this is easier to calculate the `appData` given the JSON (just hashing content using keccak)
- You are more in favor on using keccak instead of sha-256


I don't disagree with you, it's just, you are discussing a change that shouldn't be taken lightly since it has some ramifications. 

I won't fight against this change, however, if you really want to push for this, I would just ask to consider these aspects when you make this decision:
- Right now, it is deterministic. I agree your solution is easier for creating an `app-data`, but it moves some of the complexity to the READING. The reading affects also decentralization since for example TheGraph could index these documents with the current implementation (this is one of the reasons we went for it). So if we change, please just make sure we don't need to depend on a private database or is too complex to be implemented in simpler indexers.
- We should make sure this breaking change, its properly handled. This will affect the Explorer, Dune, CoW Swap, and some documentation in Notion and doc.cow.fi regarding the app-data. If you go for this, please, make sure we make things coherent and well-planned. 
- We should try to coordinate with the integrations we have and are using those codes, so they replace them with new ones. 

Some links related to documentation:
- https://www.notion.so/cownation/appData-Orders-Metadata-020d5f4e18bd45248a26ec5b16dba6df?pvs=4
- https://docs.cow.fi/cow-sdk/order-meta-data-appdata
- Readme https://github.com/cowprotocol/app-data

If I mention all this, is just not because I'm opposed to this change, quite the contrary, I just don't want to coordinate such a change or put a lot of hours into doing the changes, but if you help with this part, and make a smooth transition I like your idea 👍

**Comment by bh2smith** [comment]
Just since it has not been mentioned anywhere (it seems people are not aware of how the appData gets into Dune), we would also have to rework how we fetch content from the appHash [here in dune-sync](https://github.com/cowprotocol/dune-sync/blob/main/src/fetch/ipfs.py). This would mean (I suppose) that we would have to make possibly 2 requests everytime we want to fetch the content (one for the old format and one for the new one). alternatively we could have some sort of cutof point (since we know the "first seen block" of an appHash) where we only check the new style for all new app hashes after a certain block.

**Comment by vkgnosis** [comment]
Good summary of the involved projects, Anxo and Ben. It is true that this is a backward incompatible change, that we need to make sure is handled right in the various services. For anything trying to go from app data hash to full app data, it needs to consider both the old way of creating the CID and the new way. Either way the transformation is simple but you don't know in advance which method was used so you have to check two CIDs. We can start enforcing the new date at some point in the API so we at least know for orders created after that date that they use the new scheme. Until services are updated, if they try to retrieve new app data with the old method, they will observe it as not existing. This isn't different from orders that today already have an app data hash that doesn't map to IPFS.

I feel I haven't communicated the motivation for this change enough. In both the old and the new schema, you do something to your input json that ends up with a CID and then you store part of that CID in the app data hash. In both cases this thing you store is the hash that is included in the CID. Full CIDs are longer than 32 bytes because they store more information than just the hash. They store the CID version, the format of the data, which hash function was used, the hash. So in both cases when we go back from app data hash to CID we fill in the agreed on defaults and this can be done in a couple of lines of code that appends the hash bytes to some hardcoded bytes.

The part that is different between old and new method is what IPFS data type the CID refers to. In the old version it is a data type called unixfsv1. This data type is complex because it can handles arbitrary directory structures. It doesn't refer to just one file but many files with names, creation date, unix permissions and so on. Creating this data type is difficult. That is why it is abstracted through a javascript library. This part is *not* deterministic. It depends on the implementation details of the library. The library probably doesn't have randomness in it's implementation (but it could) but at least when the library is updated to change some default settings or the internal algorithm, the IPFS hash would change. When that happens you will no longer be able to verify old app data. You can still read it but you can't verify it. Imagine you have an app data hash that isn't already in IPFS. You then have some json that you think the hash is from and you want to verify it. This will become impossible unless you depend on exactly the old library version that was used when the hash was created.

The new method is simpler and actually deterministic because it is fully specified without relying on any library's implementation details. This is the 20 lines of Rust code I attached before. Compare this to all the lines of code needed to implement unixfsv1 in the ipfs javascript library. You will always be able to verify new app data.

Reading the data out form IPFS is simpler with the new approach too. In the old approach you need to handle the full unixfsv1 data format. For example, I could today construct an order with an app data hash, that maps to several files in IPFS. I wonder if all our IPFS related code can handle that. On the other hand with the new approach the data you read is always one blob of binary data. There are no files.

> TheGraph could index these documents with the current implementation

Both app data CIDs can be easily retrieved through for example Cloudflare IPFS gateway, the command line, or IPFS libraries. Whether the TheGraph specifically supports it, I don't know. It should be easier for them to support the new method than the old one so I hope they do. I can try finding out.

**Comment by anxolin** [comment]
Thanks for going at great length explaining it. 

I would think is deterministic as long as you pin a dependency. Where I understand you have the biggest concern is, given the same JSON two libraries might generate different CIDs.

Let me summarise my though: 
- I do agree with you, you convinced me that your approach is better
- I'm happy we change it
- Changing will bring a lot of changes in docs, our JS, and Dune. If you are happy to coordinate this part, and help, it will be more likely to happen. From my side, i feel little energy to change this in this moment TBH given all the other work we have (i will support you, but i can't drive this one)

If you push for this, just consider this:
- Cost of the change, vs win of it
- If we are sure this long term will pay off, you are right that the earlier the better

Let me do a thought experiment, what happens if we don't do it:
- True, given same content you don't have guaranteed the CID will be the same for different implementations. 
- However, given a CID you will get the right content
- So its only important that the one create the `appData`  for the order, adds the document in IPFS. So if the backend starts to create them, you can use any CID/multihash library (and there's some for RUST), this will give you the CID and that will just work for everyone. No-one will need to re-create the CID from the content, only the other way around.
- No change for Dune
- No change for the docs
- No change for depedant projects
- No change for CoW Swap, Explorer, etc.

Now, it's your call. I promise I don't try to convince you, just give you other angles so you evaluate on not just the best possible implementation, but also the most pragmatic. I actually will be happy if this ends up with a better implementation, and no friction with these ramifications

**Comment by nlordell** [comment]
> I would think is deterministic as long as you pin a dependency

My understanding is that "library" is generally about unixfsv1 and not necessarily the JS library that is computing the CID

If you `ipfs add file.json`, the CID is actually a hash of the "fs node" for that specific file and not just its content. `unixfsv1` includes a lot of additional metadata, like file permissions for example. What happens if:
- The `ipfs` command line tool adds a `file.json` with different permission bits?
- The IPFS node you are using is configured (for arguments sake, I don't know if this is possible) to use different default file permissions?
- The IPFS node you are using uses some hypothetical `unixfsv2`?

In this regard, the CID hash that is computed is only stable so long as the use of `unixfsv1` remains unchanged for uploading the documents, while as using RAW format is stable (since the CID is uniquely computed from the content of the hashed file and not additional metadata). If I understood correctly, this is not solved by just pinning JS libraries, as it may involve externalities like IPFS node configuration and implementation.

@vkgnosis, is my understanding correct?

**Comment by nlordell** [comment]
Regarding reducing breaking-change-ness of this proposal, one mitigation (that has been informally discussed but not finalized) is to do the following for orders created with only `appData` hashes:

1. Fetch app data pre-image on IPFS
2. If that fails, check a list of "app data pre-image overrides" for that app data

This way, we can continue to support integrations that use fixed app data values without requiring them to change, while still ensuring that some verified "signed" app data rules are being upheld. 

Also, this allows us to specify the `0x00..00` app data hash to some "default" app data document (as well as other integrations that use fixed app data hashes computed in non-standard ways).

**Comment by vkgnosis** [comment]
> is my understanding correct?

Pinning the current frontend ipfs library and all of it's transitive dependencies does produce deterministic CIDs. (Assuming that the library doesn't have inherent randomness.) The problem is that it is unreasonable to rely on the implementation details of all this pinned code to form the specification of the app data hashing scheme. It is also unlikely that the pinned libraries will continue to work forever. Any upgrade has a chance to break the specification.

It is also possible that IPFS nodes might stop supporting unixfsv1 when a better unixfsv2 comes out. I expect them to keep supporting the old content for a long time. The raw hashing that this proposal uses is the most basic form of storing data in IPFS, making it the format least likely to become unsupported.

**Comment by nlordell** [comment]
> Pinning the current frontend ipfs library and all of it's transitive dependencies does produce deterministic CIDs.

Doesn't this depend on how the file upload happens though? IPFS has a "add file" HTTP API:

```
curl -X POST -F file=@dist/index.html "https://ipfs.infura.io:5001/api/v0/add?pin=true"
```

If the JS library is using such an API, then it does necessarily control how the fs node is generated. Permission bits and other metadata are determined by the IPFS node in this case right? So doesn't this mean that the CID computed by the JS library may not line up with the CID of the content stored on the IPFS node (if the node uses different defauls, `unixfs` versions, etc.)?

**Comment by vkgnosis** [comment]
If that is what the JS library is doing, then yes. I did not expect it to do that.

**Comment by nlordell** [comment]
I just checked and the API call it is doing to upload and pin sends this information: 
<img width="247" alt="image" src="https://github.com/cowprotocol/services/assets/4210206/2551f553-5573-4c4a-bc6f-0d5d14fd0d52">

AFAIU, this means that how we currently do things is unstable: we do a generic "upload this file" to Pinata, so what metadata it uses for `unixfsv1` is not specified by us and that the way we compute the CID locally happens to coincide with how Pinata does it (because we happen to have the same `unixfs` defaults). (cc @anxolin)

*Disclaimer*: my understanding of IPFS internals (how a "fs node" is set up for some content) is limited to what I have read in this thread.

**Comment by alfetopito** [comment]
Not much to add.
I've been following the discussion and Valentin raises valid points.

I'm not too concerned with the (non)deterministic nature of the appData because as Anxo pointed out, as long as they can be found in IPFS, the way how it was created is not relevant IMO.

Still, I'm also not opposed to migrating to the new format.
The concern is the same as Anxo's regarding the different parts involved and the coordination required.

I'd be happy to delegate the creation/storage of appData content to the backend as it's one less thing we have to deal with on the frontend :)

**Comment by fleupold** [comment]
Trying to summarize this discussion before the meeting we have today:
- We want to support app data being solely stored on IPFS for decentralisation reasons. For those we don't really care how they are created as long as the hash can be found (and pinned/mirrored by us)
- We want to allow the client to send the pre-image to the backend to handle the IPFS uploading part. The backend will use a simpler deterministic content-only based format.

The open question is what should happen if the backend receives an order without a pre-image that it cannot find on IPFS? Should it reject the order (which would technically be a breaking change, although we can probably mitigate breaking things in practice by whitelisting the hashes existing integrations use) or should it accept it (and assume we have a plain order without any meta-fields)?

I'd really like to avoid forcing integrations such as Balancer and Defi Llama to update their logic if possible. We are already having such a hard time getting people to integrate CoW Swap and I'm worried we won't be able to convince them to make changes (there are always other more pressing priorities)
